File: backend/app/schemas/creative.py
"""Schemas for high-level creative image actions, provenance, and results."""

from enum import Enum
from typing import Any, Dict, List, Optional
from pydantic import BaseModel, Field


class CreativeActionType(str, Enum):
    TXT2IMG = "txt2img"
    IMG2IMG = "img2img"
    INPAINT = "inpaint"
    UPSCALE = "upscale"


class CreativeActionRequest(BaseModel):
    action: CreativeActionType
    prompt: str = ""
    negative_prompt: str = "low quality, blurry, deformed, bad anatomy"
    model: str = "v1-5-pruned-emaonly.safetensors"
    engine_id: str = "managed_comfyui"  # managed_comfyui, managed_webui, or cloud
    aspect_ratio: str = "1:1"  # 1:1, 16:9, 9:16, 4:3
    width: int = 512
    height: int = 512
    steps: int = 20
    cfg_scale: float = 7.0
    seed: int = -1  # -1 for random
    denoise: float = 0.75  # For img2img and inpaint (0.0 to 1.0)
    input_image_id: Optional[str] = None  # Reference asset ID for img2img / inpaint / upscale
    mask_image_id: Optional[str] = None   # Binary mask asset ID for inpaint
    upscale_factor: float = 2.0           # 2.0 or 4.0
    upscaler_name: str = "R-ESRGAN 4x+"


class GenerationProvenance(BaseModel):
    action: CreativeActionType
    prompt: str
    negative_prompt: Optional[str] = None
    model: str
    engine_id: str
    seed: int
    steps: int
    cfg_scale: float
    dimensions: str  # e.g. "512x512"
    created_at: str
    source_asset_id: Optional[str] = None
    mask_asset_id: Optional[str] = None
    execution_time_ms: Optional[float] = None


class CreativeActionResult(BaseModel):
    success: bool
    task_id: str
    asset_id: Optional[str] = None
    image_url: Optional[str] = None
    width: int = 512
    height: int = 512
    provenance: Optional[GenerationProvenance] = None
    is_cached: bool = False
    error_message: Optional[str] = None

File: frontend/src/types/creative.ts
export type CreativeActionType = 'txt2img' | 'img2img' | 'inpaint' | 'upscale' | 'txt2video' | 'img2video';

export interface GenerationProvenance {
  action: CreativeActionType;
  prompt: string;
  negative_prompt?: string;
  model: string;
  engine_id: string;
  seed: number;
  steps: number;
  cfg_scale: number;
  dimensions: string;
  created_at: string;
  source_asset_id?: string;
  mask_asset_id?: string;
  execution_time_ms?: number;
  fps?: number;
  num_frames?: number;
  duration_seconds?: number;
  motion_bucket_id?: number;
}

export interface CreativeActionRequest {
  action: CreativeActionType;
  prompt: string;
  negative_prompt?: string;
  model?: string;
  engine_id?: string;
  aspect_ratio?: string;
  width?: number;
  height?: number;
  steps?: number;
  cfg_scale?: number;
  seed?: number;
  denoise?: number;
  input_image_id?: string;
  mask_image_id?: string;
  upscale_factor?: number;
  upscaler_name?: string;
  fps?: number;
  num_frames?: number;
  motion_bucket_id?: number;
  duration_seconds?: number;
}

export interface CreativeActionResult {
  success: boolean;
  task_id: string;
  asset_id?: string;
  image_url?: string;
  video_url?: string;
  width: number;
  height: number;
  duration_seconds?: number;
  fps?: number;
  provenance?: GenerationProvenance;
  is_cached?: boolean;
  error_message?: string;
}

export interface ImageCardData {
  assetId?: string;
  imageUrl: string;
  videoUrl?: string;
  mediaType?: 'image' | 'video';
  width: number;
  height: number;
  provenance?: GenerationProvenance;
  label?: string;
}

export interface AgentActionStep {
  step_number: number;
  action: CreativeActionType;
  engine_id: string;
  model: string;
  parameters: Record<string, any>;
  description?: string;
}

export interface WorkflowNodeStage {
  stage_number: number;
  name: string;
  node_type: string;
  description: string;
}

export interface ReviewableWorkflowGraph {
  workflow_id: string;
  workflow_title: string;
  node_count: number;
  stages: WorkflowNodeStage[];
  required_nodes: string[];
  required_models: string[];
  missing_models: string[];
  is_valid: boolean;
  validation_issues: string[];
  recovery_guidance?: string;
}

export interface AgentProposal {
  id: string;
  intent: string;
  title: string;
  summary: string;
  target_engine: string;
  model: string;
  parameters: Record<string, any>;
  chain_steps: AgentActionStep[];
  workflow_graph?: ReviewableWorkflowGraph;
  estimated_calls: number;
  cost_disclaimer: string;
  explanation: string;
  requires_user_approval: boolean;
  approved: boolean;
}

export interface AgentChatMessage {
  id: string;
  role: 'user' | 'assistant' | 'system';
  content: string;
  proposal?: AgentProposal;
  created_at: string;
}

